File: apps/desktop/src/data/fixtures/inventory.ts
// Copyright (C) 2024-2026 Sjors Robroek
// SPDX-License-Identifier: AGPL-3.0-only

/**
 * Static fixture data for the Inventory ledger (spec 006).
 *
 * Types mirror the `inventory.list` contract from
 * `specs/006-inventory-library-lifecycle/contracts/inventory.list.json`.
 * These are used by the mock invoke path (VITE_USE_MOCKS=true) and vitest.
 *
 * Spec 041 FR-051 (T076, Phase 13): sessions are derived, already-confirmed
 * inventory. The review-state machine (`SessionState`, `InventorySession.state`,
 * `InventoryListRequest.filters.reviewFilter`) and the
 * `InventorySessionReviewRequest`/`InventorySessionReviewResponse` DTOs were
 * removed along with the `inventory.session.review` command they backed.
 */

// ── Contract DTOs (local inline until bindings are regenerated) ───────────────

// `mixed` removed 2026-07-03 (spec 006 iteration): Inbox single-type ingest
// (spec 041) splits mixed folders at ingest, so an inventory item is never mixed.
export type FrameType = 'light' | 'dark' | 'flat' | 'bias';
export type SourceKind =
  | 'local_disk'
  | 'external_disk'
  | 'removable'
  | 'network_share';
export type SourceState =
  | 'active'
  | 'missing'
  | 'disabled'
  | 'reconnect_required';

export interface ProvenanceSummary {
  target?: string;
  filter?: string;
  inferred?: string;
  confirmedBy?: string;
}

export interface LinkedProjectRef {
  id: string;
  name: string;
}

export interface LinkedRefs {
  projects?: LinkedProjectRef[];
  session?: string;
  calibration?: string;
}

export interface InventorySession {
  id: string;
  name: string;
  sourceId: string;
  frames: number;
  type: FrameType;
  target: string | null;
  filter: string | null;
  exposure: string | null;
  camera?: string;
  gain?: string;
  binning?: string;
  setTemp?: string;
  capturedOn?: string;
  provenance?: ProvenanceSummary;
  linked?: LinkedRefs;
}

export interface InventorySource {
  id: string;
  path: string;
  kind: SourceKind;
  state: SourceState;
  sessions: InventorySession[];
}

export interface InventoryListRequest {
  contractVersion: string;
  requestId: string;
  filters?: {
    sourceFilter?: string;
    frameFilter?: FrameType;
  };
}

export interface InventoryListResponse {
  status: 'success' | 'error';
  contractVersion: string;
  requestId: string;
  generatedAt: string;
  sources: InventorySource[];
}

// ── Fixture constant UUIDs ────────────────────────────────────────────────────

const ROOT_LOCAL = '550e8400-e29b-41d4-a716-000000000001';
const ROOT_EXTERNAL = '550e8400-e29b-41d4-a716-000000000002';

const PROJECT_NGC7000_NB = '550e8400-e29b-41d4-a716-446655440301';
const PROJECT_M31_LRGB = '550e8400-e29b-41d4-a716-446655440302';

// ── Fixture sessions ──────────────────────────────────────────────────────────

const LOCAL_SESSIONS: InventorySession[] = [
  {
    id: '550e8400-e29b-41d4-a716-446655440001',
    name: 'NGC 7000 · Ha — 2026-04-12',
    sourceId: ROOT_LOCAL,
    frames: 54,
    type: 'light',
    target: 'NGC 7000',
    filter: 'Ha',
    exposure: '300s',
    camera: 'ASI2600MM Pro',
    gain: '100',
    binning: '1×1',
    setTemp: '-10°C',
    capturedOn: '2026-04-12',
    provenance: { target: 'NGC 7000', filter: 'Ha', confirmedBy: 'user' },
    linked: {
      projects: [
        { id: PROJECT_NGC7000_NB, name: 'NGC 7000 · HOO' },
        {
          id: '550e8400-e29b-41d4-a716-446655440303',
          name: 'NGC 7000 · SHO mosaic',
        },
      ],
    },
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440002',
    name: 'NGC 7000 · OIII — 2026-04-15',
    sourceId: ROOT_LOCAL,
    frames: 22,
    type: 'light',
    target: 'NGC 7000',
    filter: 'OIII',
    exposure: '600s',
    camera: 'ASI2600MM Pro',
    gain: '100',
    binning: '1×1',
    setTemp: '-10°C',
    capturedOn: '2026-04-15',
    provenance: { target: 'NGC 7000', filter: 'OIII', confirmedBy: 'user' },
    linked: { projects: [{ id: PROJECT_NGC7000_NB, name: 'NGC 7000 · HOO' }] },
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440003',
    name: 'IC 1396 · SII — 2026-04-14',
    sourceId: ROOT_LOCAL,
    frames: 18,
    type: 'light',
    target: 'IC 1396',
    filter: 'SII',
    // 360 s exposure; GAIN=125 (narrowband mode), OFFSET=20, BIN=1×1, SET-TEMP=0°C
    // from real Poseidon-C PRO / NINA 3.1.2 header (IMAGETYP=LIGHT, XBINNING=1)
    exposure: '360s',
    camera: 'Poseidon-C PRO',
    gain: '125',
    binning: '1×1',
    setTemp: '0°C',
    capturedOn: '2026-04-14',
    provenance: { target: 'IC 1396', inferred: 'from FITS OBJECT' },
    linked: undefined,
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440006',
    name: 'M31 · L — 2026-03-28',
    sourceId: ROOT_LOCAL,
    frames: 120,
    type: 'light',
    target: 'M31',
    filter: 'L',
    // 195 s exposure; GAIN=0 (low-noise LUM mode), OFFSET=20, BIN=1×1, SET-TEMP=0°C
    // from real Poseidon-C PRO / NINA 3.1.2 header (IMAGETYP=LIGHT, XBINNING=1)
    exposure: '195s',
    camera: 'Poseidon-C PRO',
    gain: '0',
    binning: '1×1',
    setTemp: '0°C',
    capturedOn: '2026-03-28',
    provenance: {
      target: 'M31',
      filter: 'L',
      inferred: 'filter origin inferred',
    },
    linked: undefined,
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440007',
    name: 'M31 · B — 2026-03-10',
    sourceId: ROOT_LOCAL,
    frames: 35,
    type: 'light',
    target: 'M31',
    filter: 'B',
    // 100 s exposure; GAIN=0, OFFSET=20, BIN=1×1, SET-TEMP=0°C
    // from real Poseidon-C PRO / NINA 3.1.2 header (IMAGETYP=LIGHT, XBINNING=1)
    exposure: '100s',
    camera: 'Poseidon-C PRO',
    gain: '0',
    binning: '1×1',
    setTemp: '0°C',
    capturedOn: '2026-03-10',
    linked: undefined,
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440008',
    name: 'M42 · OIII — 2026-02-11',
    sourceId: ROOT_LOCAL,
    frames: 28,
    type: 'light',
    target: 'M42',
    filter: 'OIII',
    // 171 s exposure; GAIN=125 (narrowband mode), OFFSET=20, BIN=1×1, SET-TEMP=0°C
    // from real Poseidon-C PRO / NINA 3.1.2 header (IMAGETYP=LIGHT, XBINNING=1)
    exposure: '171s',
    camera: 'Poseidon-C PRO',
    gain: '125',
    binning: '1×1',
    setTemp: '0°C',
    capturedOn: '2026-02-11',
    linked: undefined,
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440009',
    name: 'M42 · Ha — 2026-02-10',
    sourceId: ROOT_LOCAL,
    frames: 45,
    type: 'light',
    target: 'M42',
    filter: 'Ha',
    // 300 s exposure; GAIN=125 (narrowband mode), OFFSET=20, BIN=1×1, SET-TEMP=0°C
    // from real Poseidon-C PRO / NINA 3.1.2 header (IMAGETYP=LIGHT, XBINNING=1)
    exposure: '300s',
    camera: 'Poseidon-C PRO',
    gain: '125',
    binning: '1×1',
    setTemp: '0°C',
    capturedOn: '2026-02-10',
    linked: undefined,
  },
];

const EXTERNAL_SESSIONS: InventorySession[] = [
  {
    id: '550e8400-e29b-41d4-a716-446655440005',
    name: 'M31 · R — 2026-03-30',
    sourceId: ROOT_EXTERNAL,
    frames: 40,
    type: 'light',
    target: 'M31',
    filter: 'R',
    // 101 s exposure; GAIN=0, OFFSET=20, BIN=1×1, SET-TEMP=0°C
    // from real Poseidon-C PRO / NINA 3.1.2 header (IMAGETYP=LIGHT, XBINNING=1)
    exposure: '101s',
    camera: 'Poseidon-C PRO',
    gain: '0',
    binning: '1×1',
    setTemp: '0°C',
    capturedOn: '2026-03-30',
    provenance: { target: 'M31', filter: 'R', confirmedBy: 'user' },
    linked: { projects: [{ id: PROJECT_M31_LRGB, name: 'M31 · LRGB' }] },
  },
  {
    id: '550e8400-e29b-41d4-a716-446655440010',
    name: 'NGC 2244 · Ha — 2026-01-20',
    sourceId: ROOT_EXTERNAL,
    frames: 62,
    type: 'light',
    target: 'NGC 2244',
    filter: 'Ha',
    // 300 s exposure; GAIN=0, OFFSET=50, BIN=1×1, SET-TEMP=0°C
    // from real ZWO ASI2600MM Pro / NINA 3.2.0 header (IMAGETYP=LIGHT, XBINNING=1, FILTER=OIII)
    exposure: '300s',
    camera: 'ZWO ASI2600MM Pro',
    gain: '0',
    binning: '1×1',
    setTemp: '0°C',
    capturedOn: '2026-01-20',
    linked: undefined,
  },
  // Calibration set on external drive
  {
    id: '550e8400-e29b-41d4-a716-446655440020',
    name: 'dark calibration — 2026-04-01',
    sourceId: ROOT_EXTERNAL,
    frames: 20,
    type: 'dark',
    target: null,
    filter: null,
    // 300 s exposure; GAIN=60, BIN=1×1 — from real DWARFIII master dark filename
    // dark_exp_300.000000_gain_60_bin_1_44C_stack_10.fits
    exposure: '300s',
    camera: 'DWARFIII',
    gain: '60',
    binning: '1×1',
    setTemp: '−44°C',
    capturedOn: '2026-04-01',
    linked: undefined,
  },
];

// ── Exported fixture data ─────────────────────────────────────────────────────

export const INVENTORY_SOURCES: InventorySource[] = [
  {
    id: ROOT_LOCAL,
    path: '/home/user/astrophotos',
    kind: 'local_disk',
    state: 'active',
    sessions: LOCAL_SESSIONS,
  },
  {
    id: ROOT_EXTERNAL,
    path: '/media/ExternalSSD/astrophotos',
    kind: 'external_disk',
    state: 'active',
    sessions: EXTERNAL_SESSIONS,
  },
];

/** Default mock response for inventory.list. */
export const INVENTORY_LIST_RESPONSE: InventoryListResponse = {
  status: 'success',
  contractVersion: '2.0.0',
  requestId: '00000000-0000-0000-0000-000000000001',
  generatedAt: '2026-06-11T00:00:00Z',
  sources: INVENTORY_SOURCES,
};

File: crates/contracts/core/src/inventory.rs
// Copyright (C) 2024-2026 Sjors Robroek
// SPDX-License-Identifier: AGPL-3.0-only

//! Inventory contract DTOs for the Tauri IPC surface (spec 006).
//!
//! These types mirror the JSON Schema contracts in
//! `specs/006-inventory-library-lifecycle/contracts/`. They are the canonical
//! Rust-side DTO boundary for `inventory.list`.
//!
//! The UI-visible `InventorySource` / `InventorySession` types are **read-only
//! projections** — no new persisted entities are introduced by spec 006.
//!
//! Spec 041 FR-051 (T076, Phase 13): sessions are derived, already-confirmed
//! inventory. The review-state machine (`InventorySessionState`) and the
//! `inventory.session.review` command that wrapped the spec-002
//! `lifecycle.transition` use case were removed; there is no longer a
//! review-state mutation on this surface. Session metadata remains editable
//! post-hoc via the inbox per-file metadata/override tables.

use serde::{Deserialize, Serialize};
use specta::Type;

use crate::sessions::SessionCalibrationMatch;

// ── FrameType ────────────────────────────────────────────────────────────────

/// Frame type for an inventory session.
/// `DarkFlat` is reserved but never returned in v1.
/// (`Mixed` removed 2026-07-03: Inbox single-type ingest — spec 041 — splits
/// mixed folders into single-type items at ingest, so a session is never mixed.)
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum InventoryFrameType {
    Light,
    Dark,
    Flat,
    Bias,
}

// ── SourceKind / SourceState ─────────────────────────────────────────────────

/// Library root media kind (refined from `LibraryRoot.kind`).
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum InventorySourceKind {
    LocalDisk,
    ExternalDisk,
    Removable,
    NetworkShare,
}

/// Library root connectivity state (mirrors `LibraryRoot.state` from spec 002).
#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize, Type)]
#[serde(rename_all = "snake_case")]
pub enum InventorySourceState {
    Active,
    Missing,
    Disabled,
    ReconnectRequired,
}

// ── Provenance / Linked ──────────────────────────────────────────────────────

/// Provenance summary for the detail drawer. MUST NOT include
/// confidence/evidence detail (spec 002 FR-006). Summary only.
#[derive(Clone, Debug, Default, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct InventoryProvenanceSummary {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filter: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub inferred: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub confirmed_by: Option<String>,
}

/// Lightweight project stub in the linked section.
#[derive(Clone, Debug, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct LinkedProjectRef {
    pub id: String,
    pub name: String,
}

/// Outbound references shown in the drawer's "Linked" section.
#[derive(Clone, Debug, Default, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct InventoryLinkedRefs {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub projects: Option<Vec<LinkedProjectRef>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub calibration: Option<String>,
}

// ── InventorySession ─────────────────────────────────────────────────────────

/// One row in the inventory ledger. Projects one `AcquisitionSession` OR one
/// `CalibrationSession` into a unified DTO.
///
/// Spec 041 FR-051: no `state` field — sessions are derived, already-confirmed
/// inventory with no review lifecycle.
#[derive(Clone, Debug, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct InventorySession {
    pub id: String,
    pub name: String,
    pub source_id: String,
    pub frames: u32,
    #[serde(rename = "type")]
    pub frame_type: InventoryFrameType,
    pub target: Option<String>,
    pub filter: Option<String>,
    pub exposure: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub camera: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gain: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub binning: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub set_temp: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub captured_on: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provenance: Option<InventoryProvenanceSummary>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub linked: Option<InventoryLinkedRefs>,
    /// The session's frame folder, relative to its source root
    /// (`source_id`'s `current_path`). The reveal action joins the root path
    /// with this so it opens the session's actual frame folder instead of the
    /// library root (#567). `None` when no frame `file_record` resolves a
    /// path (legacy/unscanned sessions) — the UI then falls back to the root.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relative_path: Option<String>,
    /// User-editable free-text notes (#773). `None` when never set.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub notes: Option<String>,
    /// Calibration masters assigned to this session (`calibration_assignment`
    /// rows), reusing the same DTO the (frontend-unused) `sessions_get`
    /// contract already carries. Empty for calibration sessions (dark/flat/
    /// bias) — assignment links a light session to its calibration masters,
    /// never the reverse — and for a light session with no assignment yet;
    /// the UI renders both as an explicit "no calibration match" state
    /// rather than omitting the section (#772).
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub calibration_matches: Vec<SessionCalibrationMatch>,
}

// ── InventorySource ──────────────────────────────────────────────────────────

/// One group header in the inventory ledger. One per `LibraryRoot` that has at
/// least one session under it.
#[derive(Clone, Debug, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct InventorySource {
    pub id: String,
    pub path: String,
    pub kind: InventorySourceKind,
    pub state: InventorySourceState,
    pub sessions: Vec<InventorySession>,
}

// ── inventory.list request / response ────────────────────────────────────────

/// Optional filters for `inventory.list`.
#[derive(Clone, Debug, Default, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct InventoryListFilters {
    /// When set, limits the response to a single `LibraryRoot`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_filter: Option<String>,
    /// When set, limits sessions to the given frame type.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frame_filter: Option<InventoryFrameType>,
    /// Maximum sessions returned per source root (default 1 000 server-side
    /// when omitted). Existing callers that omit the field keep working.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
    /// Sessions to skip before applying `limit` (0-based, per source root).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub offset: Option<u32>,
}

/// Request envelope for `inventory.list`.
#[derive(Clone, Debug, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct InventoryListRequest {
    pub contract_version: String,
    pub request_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filters: Option<InventoryListFilters>,
}

/// Successful response payload for `inventory.list`.
#[derive(Clone, Debug, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct InventoryListResponse {
    pub status: String,
    pub contract_version: String,
    pub request_id: String,
    pub generated_at: String,
    pub sources: Vec<InventorySource>,
}

// ── inventory.session.notes.update request / response (#773) ───────────────────

/// Request for `inventory.session.notes.update`. Mirrors the
/// `target.note.update` shape (spec 023 US4) — empty/whitespace-only `notes`
/// clears the field (stores NULL).
#[derive(Clone, Debug, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct SessionNotesUpdateRequest {
    pub session_id: String,
    pub notes: String,
}

/// Response for `inventory.session.notes.update`.
#[derive(Clone, Debug, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct SessionNotesUpdateResult {
    /// Notes after the update, or `null` when cleared.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub notes: Option<String>,
}
